modules/compiler/utils/source/cl_builtin_info.cpp

```diff
@@ -2982,6 +2989,451 @@ CallInst *CLBuiltinInfo::mapSyncBuiltinToMuxSyncBuiltin(
+      RecheckOpType = true;
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanUMaxExclusive;
+      break;
+    case eCLBuiltinWorkgroupScanMulInclusive:
+      RecheckOpType = true;
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanMulInclusive;
+      break;
+    case eCLBuiltinWorkgroupScanMulExclusive:
+      RecheckOpType = true;
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanMulExclusive;
+      break;
+    case eCLBuiltinWorkgroupScanAndInclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanAndInclusive;
+      break;
+    case eCLBuiltinWorkgroupScanAndExclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanAndExclusive;
+      break;
+    case eCLBuiltinWorkgroupScanOrInclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanOrInclusive;
+      break;
+    case eCLBuiltinWorkgroupScanOrExclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanOrExclusive;
+      break;
+    case eCLBuiltinWorkgroupScanXorInclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanXorInclusive;
+      break;
+    case eCLBuiltinWorkgroupScanXorExclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanXorExclusive;
+      break;
+    case eCLBuiltinWorkgroupScanLogicalAndInclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanLogicalAndInclusive;
+      break;
+    case eCLBuiltinWorkgroupScanLogicalAndExclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanLogicalAndExclusive;
+      break;
+    case eCLBuiltinWorkgroupScanLogicalOrInclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanLogicalOrInclusive;
+      break;
+    case eCLBuiltinWorkgroupScanLogicalOrExclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanLogicalOrExclusive;
+      break;
+    case eCLBuiltinWorkgroupScanLogicalXorInclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanLogicalXorInclusive;
+      break;
+    case eCLBuiltinWorkgroupScanLogicalXorExclusive:
+      MuxBuiltinID = eMuxBuiltinWorkgroupScanLogicalXorExclusive;
+      break;
+  }
+
+  if (RecheckOpType) {
+    // We've assumed (unsigned) integer operations, but we may actually have
+    // signed integer, or floating point, operations.
+    compiler::utils::NameMangler Mangler(&F->getContext());
+    SmallVector<Type *, 4> ArgumentTypes;
+    SmallVector<compiler::utils::TypeQualifiers, 4> Qualifiers;
+
+    const auto DemangledName = std::string(
+        Mangler.demangleName(F->getName(), ArgumentTypes, Qualifiers));
+
+    assert(Qualifiers.size() == 1 && ArgumentTypes.size() == 1 &&
+           "Unknown collective builtin");
+    auto &Qual = Qualifiers[0];
+
+    bool IsSignedInt = false;
+    while (!IsSignedInt && Qual.getCount()) {
+      IsSignedInt |= Qual.pop_front() == compiler::utils::eTypeQualSignedInt;
+    }
+
+    bool IsFP = ArgumentTypes[0]->isFloatingPointTy();
+    switch (MuxBuiltinID) {
+      default:
+        llvm_unreachable("unknown group operation for which to check the type");
+      case eMuxBuiltinSubgroupReduceAdd:
+        MuxBuiltinID = IsFP ? eMuxBuiltinSubgroupReduceFAdd : MuxBuiltinID;
```

Comment:
could you explain why it's an ID like eMuxBuiltinSubgroupReduceFAdd  or something generic like muxBuiltinID?

Reply:
It's because `MuxBuiltinID` has already been set to a 'base' ID in a previous step, which in this case is `eMuxBuiltinSubgroupReduceAdd`. This step is just refining that to `FAdd` if it's `IsFP`.

Reply:
It's basically to try and do less complicated logic in a previous step, because we don't want to start demangling builtins and checking types unless we have to, which is only in comparitively few situations, like add/mul/min/max operations.

Reply:
Oops, just noticed we were doing a redundant demangling in all cases anyway. I've removed that now

Reply:
I think just a comment to explain why we set it to the generic version as it's not obvious.

Reply:
done